### ISSUE TITLE
Difficulties in reproducing the GermEval14 NER model

### ISSUE THREAD

**Issue by mrshu** [nlp-uoregon/trankit#7]
Hi again @minhvannguyen,

I am sorry to bother you once again but I was wondering whether you could provide a bit more information on how might one reproduce the `trankit` results on GermEval14, which are presented [in the `trankit` paper](https://arxiv.org/pdf/2101.03289.pdf).

Baed on your suggestion in #6 I tried to train a `trankit`-based NER model on the GermEval14 data by directly passing it to  `trankit.TPipeline`. You can find the (very simple) code that sets up the environment, prepares the data and trains the model in the [following Colab](https://colab.research.google.com/drive/1sgU0U42c1ipn6QbskFcRDs_tdvQgK7Gf).

In the paper, Table 3 reports the test F1 score on this dataset at **86.9** but even after running over 80 training epochs, the best dev F1 score I managed to receive was on **1.74** and it does not seem like the evaluation on the test set would produce vastly different results.

Hence, my preliminary confusion is that I must be doing something wrong. One of the first suspects would be random seeds but those seems to be fixed as we can see in the snippet below:
https://github.com/nlp-uoregon/trankit/blob/b7e4a3bc25d564b3b2870a2b03a5aa4fc9a38c9a/trankit/tpipeline.py#L112-L119

I was therefore wondering whether you could immediately see what I am doing wrong here, or generally provide some pointers that could be helpful in reproducing the results listed in the paper.

Thanks!

**Comment by minhvannguyen** [comment]
Hi @mrshu,
I'm sorry for not letting you know in the [previous discussion](https://github.com/nlp-uoregon/trankit/issues/6) that the GermEval14 dataset needs a preprocessing step for the BIO annotations.  Below is the lines 78-92 in the training data (i.e., NER-de-train.tsv) of the GermEval14:
```
#   stern.de vom 21.03.2006 [2006-03-21]
1   Bayern  B-ORG   B-LOC
2   München I-ORG   B-LOC
3   ist O   O
4   wieder  O   O
5   alleiniger  O   O
6   Top-    O   O
7   Favorit O   O
8   auf O   O
9   den O   O
10  Gewinn  O   O
11  der O   O
12  deutschen   B-LOCderiv  O
13  Fußball-Meisterschaft   O   O
14  .   O   O
```
If you look at token #12, you can see it's annotated as 'B-LOCderiv'. We need to reduce this tag to 'B-LOC' to have the final annotation for this token. To do this reduction, the following snippet can be used:
```
import re

original_tag = 'B-LOCderiv'
reduced_tag = re.findall('[-A-Z]+', original_tag)[0]
```
Please let me know if this can solve the problem.
Thanks

**Comment by mrshu** [comment]
Thanks a lot for your suggestion @minhvannguyen -- thanks to the preprocessing step you suggested, I was able to get the dev F1 score to **87.03**, which is way better than what it looked like before :slightly_frowning_face:!

I would also very much like to evaluate the resulting model on the test set. When I called `trankit.verify_customized_pipeline` as outlined in the [Loading](https://trankit.readthedocs.io/en/latest/training.html#loading) section, I received the following warnings. 

```
Missing ./save_dir_filtered/customized-ner/customized-ner.tagger.mdl
Missing ./save_dir_filtered/customized-ner/customized-ner.vocabs.json
Missing ./save_dir_filtered/customized-ner/customized-ner_lemmatizer.pt
Customized pipeline is not ready to use!
Please consider the missing files above.
```

Furthermore, when I then tried to use the trained model in a `Pipeline`, I was met with the following error:
![Screenshot from 2021-02-08 16-21-19](https://user-images.githubusercontent.com/461491/107239981-b0f54700-6a29-11eb-9ec1-734a4437ccfc.png)

Looking at the code of the `Pipeline` (below), I am not sure how can I make it use the model that was just trained.

https://github.com/nlp-uoregon/trankit/blob/b7e4a3bc25d564b3b2870a2b03a5aa4fc9a38c9a/trankit/pipeline.py#L42 

Is there anything you could suggest I do so that the trained model could be tested out on the `test` part of the dataset? My current code/thinking can be found in the following [Colab](https://colab.research.google.com/drive/1sgU0U42c1ipn6QbskFcRDs_tdvQgK7Gf).

As I said previously, once I get it to work, I'd be happy to contribute this notebook to the project, so that it's obvious how could it be used by others.

Thanks again!

**Comment by minhvannguyen** [comment]
Hi @mrshu,
We're glad to hear that you can train your NER model on GermEval14. 

Regarding the warning returned by the `trankit.verify_customized_pipeline` function, this is due to the fact that you haven't trained the tokenizer, the tagger, and the lemmatizer for your custom pipeline.  This is required as specified by the [`customized-ner`](https://trankit.readthedocs.io/en/latest/training.html#building-a-customized-pipeline) category. As the verification step is not successful, the loading of the custom pipeline is executed with the error you showed above. 

To fix this, you can download the pretrained models for German [here](http://nlp.uoregon.edu/download/trankit/german.zip) and execute the following commands:
```
unzip german.zip
cp -a german/german.tagger.mdl  ./save_dir_filtered/customized-ner/customized-ner.tagger.mdl
cp -a german/german.vocabs.json  ./save_dir_filtered/customized-ner/customized-ner.vocabs.json
cp -a german/german_lemmatizer.pt  ./save_dir_filtered/customized-ner/customized-ner_lemmatizer.pt
```
After this, you can try the verification step again, and then we think the pipeline loading will be ok. We will work on this soon to provide an easier way to do this evaluation.

About the tutorial notebook, thank you for doing this. Please let us know when you complete your colab.
Thanks

**Comment by mrshu** [comment]
Thanks @minhvannguyen, that makes a lot of sense -- I'll be sure to use the pretrained models.

To make my question a bit more clearer, I only wanted to run a procedure similar to the one in the  `_evaluate_ner` function, just on the test set

https://github.com/nlp-uoregon/trankit/blob/87fd77ed6b0f27e9fd41e691edab5a9c75fadf07/trankit/tpipeline.py#L626-L643

Perhaps it may make sense to expose a function like that in the future versions of `trankit`?

I will certainly let you know once I finish the notebook.

Thanks again!

**Comment by mrshu** [comment]
@minhvannguyen Thanks to your suggestions, I believe I've been able to evaluate the model on the `test` set of the GermanEval 2014 -- the resulting test F1 score was **86.6**, which is quite close to the **86.9** reported in the paper.

Please find the Colab notebook on the link below -- feel free to send your suggestions, I'd be happy to incorporate them
https://colab.research.google.com/drive/1sgU0U42c1ipn6QbskFcRDs_tdvQgK7Gf

Finally, thanks a ton for all your help here. I believe this issue has already been resolved, so feel free to close it.

**Comment by minhvannguyen** [comment]
Hi @mrshu,
Thanks for your effort on this. 
It's great that you can train and evaluate your own NER model. As for your suggestion on the `_eval_ner` function, we will make it easier to use in our next versions. About your colab tutorial, we will add it to our github repo and update the docs page accordingly. 
Thank you!
